MongoDB Example (#15029)
vercel/next.js: examples/with-mongodb/README.md

```bash
cp .env.local.example .env.local
```

Set each variable on `.env.local`:

- `MONGODB_URI` - Your MongoDB connection string. If you are using [MongoDB Atlas](https://mongodb.com/atlas) you can find this by clicking the "Connect" button for your cluster.
- `MONGODB_DB` - The name of the database you want to use.

### Start Up Development Server

Navigate to your project directory and run either:

```bash
npm install
npm run dev
# or
yarn
yarn dev
```

Open [http://localhost:3000](http://localhost:3000) with your browser to see the result.

Navigate to [http://localhost:3000/api/hello](http://localhost:3000/api/hello) to test if your connection to MongoDB was successful. If it was you should see a response that looks like this:

```json
{
  "ns": "test.test",
  "size": 0,
  "count": 0,
  "storageSize": 0,
  "nindexes": 0,
  "totalIndexSize": 0,
  "indexDetails": {},
  "indexSizes": {},
  "scaleFactor": 1,
  "ok": 1,
  "$clusterTime": {
    "clusterTime": "6847589513905045623",
    "signature": {
      "hash": "fgIxihPon+Kg7i2a36XDddx3OOo=",
      "keyId": "6830955492114694147"
    }
  },
  "operationTime": "6847589513905045623"
}

> ```

Reviewer: Let's remove this, the proper way to test if the app is working should be by visiting `http://localhost:3000`

You can start editing the page by modifying `pages/index.js`. The page auto-updates as you edit the file.

## Deploy on Vercel

The easiest way to deploy your Next.js app is to use the [Vercel Platform](https://vercel.com/import?utm_medium=default-template&filter=next.js&utm_source=create-next-app&utm_campaign=create-next-app-readme) from the creators of Next.js.

Check out our [Next.js deployment documentation](https://nextjs.org/docs/deployment) for more details.

## Learn More

To learn more about Next.js, take a look at the following resources:

- [MongoDB Atlas](https://mongodb.com/atlas)

> - [MongoDB Documentation](https://docs.mongodb.com/)

Reviewer: Move these 2 links to the top of the readme (under the main description), as they are useful resources about mongodb, then remove this entire section.

## Example app using MongoDB

[MongoDB](https://www.mongodb.com/) is a general purpose, document-based, distributed database built for modern application developers and for the cloud era. This example will show you how to connect to and use MongoDB as your backend for your Next.js app.

If you want to learn more about MongoDB, visit the following pages:

- [MongoDB Atlas](https://mongodb.com/atlas)
- [MongoDB Documentation](https://docs.mongodb.com/)

## Deploy your own

Once you have access to the environment variables you'll need, deploy the example using [Vercel](https://vercel.com?utm_source=github&utm_medium=readme&utm_campaign=next-example):

[![Deploy with Vercel](https://vercel.com/button)](https://vercel.com/import/git?c=1&s=https://github.com/vercel/next.js/tree/canary/examples/with-mongodb&env=MONGODB_URI,MONGODB_DB&envDescription=Required%20to%20connect%20the%20app%20with%20MongoDB)

## How to use

> Using `create-next-app`
> 
> Execute `create-next-app` with npm or Yarn to bootstrap the example:

Reviewer: ```suggestion
### Using `create-next-app`

Execute [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app) with [npm](https://docs.npmjs.com/cli/init) or [Yarn](https://yarnpkg.com/lang/en/docs/cli/create/) to bootstrap the example:
```

```bash
npx create-next-app --example with-mongodb with-mongodb
# or
yarn create next-app --example with-mongodb with-mongodb
```

## Configuration

### Set up a MongoDB database

Set up a MongoDB database either locally or with [MongoDB Atlas for free](https://mongodb.com/atlas).

### Set up environment variables

Copy the `env.local.example` file in this directory to `.env.local` (which will be ignored by Git):

```bash
cp .env.local.example .env.local
```

Set each variable on `.env.local`:

- `MONGODB_URI` - Your MongoDB connection string. If you are using [MongoDB Atlas](https://mongodb.com/atlas) you can find this by clicking the "Connect" button for your cluster.
- `MONGODB_DB` - The name of the MongoDB database you want to use.

> ### Start Up Development Server

Reviewer: ```suggestion
### Run Next.js in development mode
```

## Example app using MongoDB

> [MongoDB](https://mongodb.com/atlas) is a general purpose, document-based, distributed database built for modern application developers and for the cloud era. This example will show you how to connect to and use MongoDB as your backend for your Next.js app.
> 
> - [MongoDB Atlas](https://mongodb.com/atlas)
> - [MongoDB Documentation](https://docs.mongodb.com/)

Reviewer: ```suggestion
[MongoDB](https://www.mongodb.com/) is a general purpose, document-based, distributed database built for modern application developers and for the cloud era. This example will show you how to connect to and use MongoDB as your backend for your Next.js app.

If you want to learn more about MongoDB, visit the following pages:

- [MongoDB Atlas](https://mongodb.com/atlas)
- [MongoDB Documentation](https://docs.mongodb.com/)
```

> ## Demo
> 

Reviewer: ```suggestion
```

[![Deploy with Vercel](https://vercel.com/button)](https://vercel.com/import/git?c=1&s=https://github.com/vercel/next.js/tree/canary/examples/with-mongodb&env=MONGODB_URI,MONGODB_DB&envDescription=Required%20to%20connect%20the%20app%20with%20MongoDB)

## How to use

Using `create-next-app`

Execute `create-next-app` with npm or Yarn to bootstrap the example:

```bash
npx create-next-app --example with-mongodb with-mongodb
# or
yarn create next-app --example with-mongodb with-mongodb
```

## Configuration

### Set up a MongoDB database

Set up a MongoDB database either locally or with [MongoDB Atlas for free](https://mongodb.com/atlas).

### Set up environment variables

Copy the `env.local.example` file in this directory to `.env.local` (which will be ignored by Git):

```bash
cp .env.local.example .env.local
```

Set each variable on `.env.local`:

- `MONGODB_URI` - Your MongoDB connection string. If you are using [MongoDB Atlas](https://mongodb.com/atlas) you can find this by clicking the "Connect" button for your cluster.
- `MONGODB_DB` - The name of the MongoDB database you want to use.

### Start Up Development Server

> Navigate to your project directory and run either:
> 
> ```bash
> npm install
> npm run dev
> # or
> yarn
> yarn dev
> ```
> 
> Open [http://localhost:3000](http://localhost:3000) with your browser to see the result.

Reviewer: This aligns more with our other examples.

````suggestion
```bash
npm install
npm run dev

# or

yarn install
yarn dev
```

Your app should be up and running on [http://localhost:3000](http://localhost:3000)! If it doesn't work, post on [GitHub discussions](https://github.com/zeit/next.js/discussions).
````

You will either see a message stating "You are connected to MongoDB" or "You are NOT connected to MongoDB". Ensure that you have provided the correct `MONGODB_URI` and `MONGODB_DB` environment variables.

When you are successfully connected, you can refer to the [MongoDB Node.js Driver docs](https://mongodb.github.io/node-mongodb-native/3.4/tutorials/collections/) for further instructions on how to query your database.

> For a super quick example you can use the `db` variable that exported from the `mongodb.js` util file to insert data by running:
> 
> ```js
> db.collection('YOUR-COLLECTION-NAME').insertOne({ hello: 'world' })
> ```
> 
> Additionally, you can query the data by running:
> 
> ```js
> db.collection('YOUR-COLLECTION-NAME').find().toArray()
> ```
> 
> You can start editing the page by modifying `pages/index.js`. The page auto-updates as you edit the file.
> 

Reviewer: This is not really related to the example but to documentation of the mongo driver, which is better featured in their official docs.

```suggestion
```

> ## Deploy on Vercel
> 
> The easiest way to deploy your Next.js app is to use the [Vercel Platform](https://vercel.com/import?utm_medium=default-template&filter=next.js&utm_source=create-next-app&utm_campaign=create-next-app-readme) from the creators of Next.js.
> 
> Check out our [Next.js deployment documentation](https://nextjs.org/docs/deployment) for more details.

Reviewer: ```suggestion
## Deploy on Vercel

You can deploy this app to the cloud with [Vercel](https://vercel.com?utm_source=github&utm_medium=readme&utm_campaign=next-example) ([Documentation](https://nextjs.org/docs/deployment)).

#### Deploy Your Local Project

To deploy your local project to Vercel, push it to GitHub/GitLab/Bitbucket and [import to Vercel](https://vercel.com/import/git?utm_source=github&utm_medium=readme&utm_campaign=next-example).

**Important**: When you import your project on Vercel, make sure to click on **Environment Variables** and set them to match your `.env.local` file.

#### Deploy from Our Template

Alternatively, you can deploy using our template by clicking on the Deploy button below.

[![Deploy with Vercel](https://vercel.com/button)](https://vercel.com/import/git?c=1&s=https://github.com/vercel/next.js/tree/canary/examples/with-mongodb&env=MONGODB_URI,MONGODB_DB&envDescription=Required%20to%20connect%20the%20app%20with%20MongoDB)
```
